text/shared/src/main/scala/fs2/data/text/CharLikeChunks.scala

```diff
@@ -69,103 +74,112 @@ trait CharLikeChunks[F[_], In] {
-      ctx.idx = 0
-      ctx.rest = tl
-      ctx
-    })
+private[text] abstract class CharArrayBuffer[F[_], T] extends AsCharBuffer[F, T] {
+
+  type Context = CharArrayContext[F, T]
+
+  def create(s: Stream[F, T]): Context = new Context(Array.empty, 0, s, 0)
+
+  def needsPull(ctx: Context): Boolean =
+    ctx.idx >= ctx.chunk.length
 
   def advance(ctx: Context): Context = {
     ctx.idx += 1
     ctx
   }
 
-  def current(ctx: CharContext): Char =
+  def current(ctx: Context): Char =
     ctx.chunk(ctx.idx)
 
+  override def mark(ctx: Context): Unit =
+    ctx.mark = ctx.idx
+
+  override def appendMarked(ctx: Context, acc: StringBuilder): Unit =
+    acc.appendAll(ctx.chunk, ctx.mark, ctx.idx - ctx.mark)
+
 }
 
-private class CharLikeStringChunks[F[_]] extends CharLikeChunks[F, String] {
-  class StringContext(var string: String, var sidx: Int, var rest: Stream[F, String])
+private[text] class CharLikeCharChunks[F[_]] extends CharArrayBuffer[F, Char] {
 
-  type Context = StringContext
+  def pullNext(ctx: Context): Pull[F, Nothing, Option[Context]] =
+    ctx.rest.pull.uncons.map(_.map { case (hd, tl) =>
+      ctx.chunk = hd.toArray
+      ctx.idx = 0
+      ctx.rest = tl
+      ctx.mark = 0
+      ctx
+    })
 
-  def create(s: Stream[F, String]): Context =
-    new StringContext("", 0, s)
+}
 
-  def needsPull(ctx: StringContext): Boolean =
-    ctx.sidx >= ctx.string.size
+private[text] class CharLikeStringChunks[F[_]] extends CharArrayBuffer[F, String] {
 
-  def pullNext(ctx: StringContext): Pull[F, Nothing, Option[StringContext]] =
+  def pullNext(ctx: Context): Pull[F, Nothing, Option[Context]] =
     ctx.rest.pull.uncons1.map(_.map { case (hd, tl) =>
-      ctx.string = hd
-      ctx.sidx = 0
+      ctx.chunk = hd.toCharArray()
+      ctx.idx = 0
       ctx.rest = tl
+      ctx.mark = 0
       ctx
     })
 
-  def advance(ctx: Context): Context = {
-    ctx.sidx += 1
-    ctx
-  }
-
-  def current(ctx: StringContext): Char =
-    ctx.string(ctx.sidx)
-
 }
 
 // BEWARE: this implementation only works for single-byte encodings, do not use this for utf-8 for instance
-private class CharLikeSingleByteChunks[F[_]](charset: Charset) extends CharLikeChunks[F, Byte] {
-
-  class ByteContext(var chunk: String, var idx: Int, var rest: Stream[F, Byte])
-
-  type Context = ByteContext
-
-  def create(s: Stream[F, Byte]): Context =
-    new ByteContext("", 0, s)
-
-  def needsPull(ctx: Context): Boolean =
-    ctx.idx >= ctx.chunk.size
+private[text] class CharLikeSingleByteChunks[F[_]](charset: Charset) extends CharArrayBuffer[F, Byte] {
 
   def pullNext(ctx: Context): Pull[F, Nothing, Option[Context]] =
     ctx.rest.pull.uncons.map(_.map { case (hd, tl) =>
       // This code is pure. The constructor replaces malformed input
       // with the charset default replacement string, so this never throws
-      ctx.chunk = new String(hd.toArray[Byte], charset)
+      ctx.chunk = new String(hd.toArray[Byte], charset).toCharArray()
```

Comment:
Would require benchmarking, but I could imagine that `charset.decode(hd.toByteBuffer).array()` is faster. Or maybe one even wants to replace `Array[Char]` with `CharBuffer`?

Reply:
I think I will eventually switch to `CharBuffer`, I don't know why I didn't yet. I will give it a try.

Reply:
After trying a bit various things it requires more investigations on my side. I will merge this PR, which lays out the ground work, and will open another one with `CharBuffer` if it is worth it. Thanks for pointing this out.